From 9ace85874077adaf3217621bf11583f0c90e8cb6 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 26 Jun 2014 15:33:19 -0400
Subject: [fs] nfs: Minor optimisation in get_layout_by_fh_locked()

Message-id: <1403796815-28580-14-git-send-email-steved@redhat.com>
Patchwork-id: 84846
O-Subject: [RHEL7.1 PATCH 13/29] NFSv4.1: Minor optimisation in get_layout_by_fh_locked()
Bugzilla: 1111170
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

If the filehandles match, but the igrab() fails, or the layout is
freed before we can get it, then just return NULL.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upsteam Id: 9a7fe9e8900baad5f6643000ea48b91aee895165
Bugzilla: 1111170
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/callback_proc.c b/fs/nfs/callback_proc.c
index 570c8a1..41db525 100644
--- a/fs/nfs/callback_proc.c
+++ b/fs/nfs/callback_proc.c
@@ -127,13 +127,13 @@ static struct pnfs_layout_hdr * get_layout_by_fh_locked(struct nfs_client *clp,
     continue;
    ino = igrab(lo->plh_inode);
    if (!ino)
-    continue;
+    break;
    spin_lock(&ino->i_lock);
    /* Is this layout in the process of being freed? */
    if (NFS_I(ino)->layout != lo) {
     spin_unlock(&ino->i_lock);
     iput(ino);
-    continue;
+    break;
    }
    pnfs_get_layout_hdr(lo);
    spin_unlock(&ino->i_lock);
-- 
1.7.1